mailpoet/lib/WooCommerce/Helper.php

```diff
@@ -223,4 +223,23 @@ public function getLatestCoupon(): ?string {
 
     return $coupon ? $coupon->post_title : null; // phpcs:ignore Squiz.NamingConventions.ValidVariableName.MemberNotCamelCaps
   }
+
+  /**
+   * Check whether the current request is processing a WooCommerce checkout.
+   * Works for both the normal checkout and the block checkout.
+   *
+   * This solution is not ideal, but I cheched with a few WooCommerce developers,
+   * and it is what they suggested. There is no helper function provided by Woo
+   * for this.
+   *
+   * @return bool
+   */
+  public function isCheckoutRequest(): bool {
+    $requestUri = !empty($_SERVER['REQUEST_URI']) ? sanitize_text_field(wp_unslash($_SERVER['REQUEST_URI'])) : '';
+    $isRegularCheckout = is_checkout();
+    $isBlockCheckout = WC()->is_rest_api_request()
```

Comment:
The way that I'm using to check if it is the regular checkout and the block checkout is a bit hacky but it is the best I could find. I also discussed a bit with Woo core developers and they didn't have a better suggestion.

`is_checkout()` is meant to return true when the page being displayed is the checkout page. It is not meant to return true when processing the checkout request. But since the request is made to the checkout page, it works.

Regarding the block checkout, I could not find any WP or Woo method that would work, so I ended up with this approach that relies on the path of the checkout endpoint, which is a brittle solution. At least, a test is likely to break if the endpoint changes in the future (for example, a new version of the store API).

Reply:
Did you explore checking to see if certain actions have fired? I was poking around at this a bit and was wondering if we could check for `did_action('woocommerce_store_api_checkout_order_processed')`. In my testing it returns `1` at this point in the code when checking out via the block checkout, and here's a comment from the [source code](https://github.com/woocommerce/woocommerce-blocks/blob/cde9117c9d57fe9a4c81501214d837e984a6c76a/src/StoreApi/Routes/V1/Checkout.php#L250-L266) (the second point specifically):

> 		 * This is similar to existing core hook woocommerce_checkout_order_processed. We're using a new action:
>		 * - To keep the interface focused (only pass $order, not passing request data).
>		 * - This also explicitly indicates these orders are from checkout block/StoreAPI.

Reply:
I haven't considered using `did_action('woocommerce_store_api_checkout_order_processed')`. Thanks for mentioning it.

Initially, I thought that I preferred the approach that you suggested, and I implemented it. I ended up using `woocommerce_store_api_checkout_update_order_from_request` instead of `woocommerce_store_api_checkout_order_processed` as the latter runs after the MailPoet code to update the subscriber.

But then I realized that if we decide to use this action, `isCheckoutRequest()` will only work if called after a specific WooCommerce action is executed, and it might cause some unexpected behavior if, in the future, we decide to call it before the action is executed in a different part of our code. The current solution seems more generic to me and not significantly more brittle as WooCommerce could deprecate the actions that we are considering using.

So, I'm torn in terms of which approach is best. What do you think?

Reply:
> it might cause some unexpected behavior if, in the future, we decide to call it before the action is executed in a different part of our code.

This is a very good point, and I appreciate the amount of thought you've put into this! I agree that it's better to go with the current solution for the reason you mention.

When I was looking at this I thought it might be nice if we could access the current `WP_REST_Request` and check for the route it's executing instead of having to rely on `$_SERVER` values, but as far as I can tell there's not a great way to achieve this, certainly not any way that would be an improvement over your current solution.

Reply:
Yeah, I also considered using the current `WP_Rest_Request` but came to the same conclusion that you did. Anyway, thanks for considering and suggesting other approaches.